VACMS-14205 Fix profile required fields. #14026

swirtSJW opened the pull request:
## Description

Closes #14205 

relates to #12187

## Testing done


## Screenshots


## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As content editor edgar.cruse 
1. visit https://pr14233-hghtp6e3u8qmoo00kebzk48sppmv1isn.ci.cms.va.gov/tuscaloosa-health-care/staff-profiles/james-hardemon
2. Edit the profile
3. Check the box for "Create profile page with biography"
4. delete any values in the First sentence and Body text
5. **UNCheck** the box for "Create profile page with biography"
6. Save the page with a revision log message
   - [x] Validate that it saves without error.
7. Edit the node.
8. Check the box for "Create profile page with biography"
9. Save the page with a revision log message
   - [x] Validate save is prevented with a notificationpointing to "First sentence" field (HTML 5 validation happens before save even occurs)
10. Add content to the "First sentence" field
11. Save the page with a revision log message
   - [x] Validate save is prevented with a notification pointing to "First sentence" field  (This is prevented after a true save attempt, not HTML5 pre-save)  **Note:** The jump link from the error report at the top of the page does not jump to the error.  This is a core bug and is not specific to this issue.
12. Add content to the "Body Text" field
13. Save the page with a revision log message
   - [x] Validate that it saves correctly.
14.  Edit the node.
15. **UNCheck** the box for "Create profile page with biography"
16. Save the page with a revision log message
   - [x] Validate that it saves correctly.



### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

swirtSJW commented:
Commands I used to restore the files from before #14233
```
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/js/set_body_to_required.es6.js
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/js/set_body_to_required.js
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php
git restore -s e4eb78311 -SW -- docroot/modules/custom/va_gov_backend/va_gov_backend.libraries.yml
git restore -s e4eb78311 -SW -- tests/behat/features/perms.feature
```

swirtSJW commented on docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:
```diff
@@ -61,8 +60,6 @@ public function __construct(
   public static function getSubscribedEvents(): array {
     return [
       'hook_event_dispatcher.form_node_centralized_content_edit_form.alter' => 'alterCentralizedContentNodeForm',
-      'hook_event_dispatcher.form_node_person_profile_edit_form.alter' => 'alterStaffProfileNodeForm',
```
This is all getting moved to the va_gov_profile module.

swirtSJW commented on docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:
```diff
@@ -125,67 +122,6 @@ private function trimNodeTitleWhitespace(NodeInterface $node) {
     $node->setTitle($title);
   }
 
-  /**
```
This is all getting moved to the va_gov_profile module.

swirtSJW commented on config/sync/field.field.node.person_profile.field_body.yml:
```diff
@@ -8,16 +8,19 @@ dependencies:
 label: 'Body text'
 description: 'Add everything else from the second sentence on. Don''t include the first sentence in this field.'
-required: true
+required: false
```
Losing this here so we can handle it in Constraint elsewhere.

edmund-dunn replied:
Like this idea!

swirtSJW commented:
Odd failure on behat.  Looking into it

```
Feature: Permissions

  @perms @content_editor
  Scenario Outline: Content editor can edit - edit test indicates save perm is true. # features/perms.feature:5
      """
      We may want to extend this test for other roles
      """
    Given I am logged in as a user with the <role> role                              # Drupal\DrupalExtension\Context\DrupalContext::assertAuthenticatedByRole()
    Then I should be able to edit an <type>                                          # Drupal\DrupalExtension\Context\DrupalContext::assertEditNodeOfType()

    Examples:
      | role             | type                               |
      | "content_editor" | "page"                             |
      | "content_editor" | "landing_page"                     |
      | "content_editor" | "health_care_region_detail_page"   |
      | "content_editor" | "event"                            |
      | "content_editor" | "event_listing"                    |
      | "content_editor" | "health_care_local_facility"       |
      | "content_editor" | "health_care_region_page"          |
      | "content_editor" | "health_care_local_health_service" |
      | "content_editor" | "press_release"                    |
      | "content_editor" | "office"                           |
      | "content_editor" | "outreach_asset"                   |
      | "content_editor" | "publication_listing"              |
      | "content_editor" | "person_profile"                   |
        Failed step: Then I should be able to edit an "person_profile"
        SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'title' cannot be null: INSERT INTO "node_field_data" ("nid", "vid", "type", "langcode", "status", "title", "uid", "created", "changed", "promote", "sticky", "default_langcode", "revision_translation_affected", "content_translation_source", "content_translation_outdated") VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2, :db_insert_placeholder_3, :db_insert_placeholder_4, :db_insert_placeholder_5, :db_insert_placeholder_6, :db_insert_placeholder_7, :db_insert_placeholder_8, :db_insert_placeholder_9, :db_insert_placeholder_10, :db_insert_placeholder_11, :db_insert_placeholder_12, :db_insert_placeholder_13, :db_insert_placeholder_14); Array
        (
            [:db_insert_placeholder_0] => 57846
            [:db_insert_placeholder_1] => 787338
            [:db_insert_placeholder_2] => person_profile
            [:db_insert_placeholder_3] => en
            [:db_insert_placeholder_4] => 0
            [:db_insert_placeholder_5] => 
            [:db_insert_placeholder_6] => 4621
            [:db_insert_placeholder_7] => 1688661931
            [:db_insert_placeholder_8] => 1688661931
            [:db_insert_placeholder_9] => 0
            [:db_insert_placeholder_10] => 0
            [:db_insert_placeholder_11] => 1
            [:db_insert_placeholder_12] => 
            [:db_insert_placeholder_13] => und
            [:db_insert_placeholder_14] => 0
        )
         (Drupal\Core\Entity\EntityStorageException)

```

swirtSJW commented:
Code is working and locked in.  Re-working the angry tests right now.  It should not shift the code.

omahane reviewed (commented):
I am not sure why the error is showing twice:
<img width="891" alt="Screenshot 2023-07-06 at 1 16 54 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/766573/8077caa4-7483-4741-8d33-d06ac2a8a6c5">

swirtSJW commented:
> I am not sure why the error is showing twice: 

I am pretty sure that is because of re-writing of the field ids that CKeditor does when it takes over the value part of the fields.   However I can poke at that to test my assumption.

swirtSJW commented:
Adding @ndouglas for eyes on the test addition and the new cypress step function for detecting html 5 errors.

swirtSJW commented:
Odd failure on phpunit
```
There were 4 failures:

1) tests\phpunit\Content\TrimNodeTitleWhitespaceTest::testConfirmExpectedOutput with data set #1 ('This had trailing spaces    ', 'This had trailing spaces')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'This had trailing spaces'
+'This had trailing spaces    '
```

swirtSJW commented:
Well at least cypress passed :)
![image](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/5752113/8c4bf37d-2cc1-4655-ad30-14a564e9b82c)

swirtSJW commented:
I figured out the odd failure.  It was a collision of eventsubscribers and the new one was wiping out the old one.  That was resolved with 
![image](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/5752113/705ca822-cdd7-4a95-8c2e-871b1978134b)

omahane reviewed (approved):
This is great! Works like a charm.

I'm blown away by what it took to get this back to working.

ndouglas commented on docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:
```diff
@@ -0,0 +1,134 @@
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * VA.gov Generic Entity Event Subscriber. Do only multi-product stuff here.
```
Is this comment accurate?

swirtSJW replied:
Bah.   No... soiled in copypasta.   :(

swirtSJW replied:
Fixed

ndouglas commented on docroot/modules/custom/va_gov_profile/src/EventSubscriber/EntityEventSubscriber.php:
```diff
@@ -0,0 +1,134 @@
+    $form['#attached']['library'][] = 'va_gov_profile/set_body_to_required';
+    $selector = ':input[name="field_complete_biography_create[value]"]';
+    $form['field_intro_text']['widget'][0]['value']['#states'] = [
+      'required' => [
+          [$selector => ['checked' => TRUE]],
+      ],
+      'visible' => [
+            [$selector => ['checked' => TRUE]],
+      ],
```
Indenting here looks messed up. Surprised this isn't being caught by tests/hooks.

swirtSJW replied:
Good catch... tests aren't catching it nor is pre-commit hooks nor my ide.   Wierd.

swirtSJW replied:
fixed

ndouglas commented on tests/cypress/integration/common/i_save_the_node.js:
```diff
@@ -7,3 +7,14 @@ Given("I save the node", () => {
     cy.wrap(pagePath).as("pagePath");
   });
 });
+
+Given(
+  "I am prevented from saving the node by {string} {string} with error {string}",
+  (fieldType, selector, html5Error) => {
+    cy.get("form.node-form input#edit-submit").click({ force: true });
+    cy.get((fieldType += ":invalid")).should("have.length", 1);
+    cy.get(selector).then(($input) => {
+      expect($input[0].validationMessage).to.eq(html5Error);
+    });
+  }
+);
```
Nice!

swirtSJW commented:
> I'm blown away by what it took to get this back to working.

Wellllll.... there was a bit easier way to get it back to working, but it was still as fragile as it had been and would have likely failed on occasion, so this was the way to make it more reliable in both function and data integrity.

ndouglas commented:
>I am not sure why the error is showing twice:

That's [this](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12415), right? 

I created a patch [here](https://www.drupal.org/project/drupal/issues/3229493) for that issue, but I think it's stopped working :weary:

ndouglas reviewed (dismissed):
Beautiful.

ndouglas commented on docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraintValidator.php:
I think this is a really good solution to the larger problem.

swirtSJW commented on docroot/modules/custom/va_gov_profile/src/Plugin/Validation/Constraint/PageRequiredFieldsConstraint.php:
```diff
@@ -0,0 +1,25 @@
+   *
+   * @var string
+   */
+  public $message = "In order to use '@checkbox' a profile page with biography, the @empty_fields must not be empty.";
```
Good call, the class name is very specific, and the only use for it is not.  I will adjust the class name.

swirtSJW replied:
Fixed

swirtSJW commented:
> > I am not sure why the error is showing twice:
> 
> That's [this](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/12415), right?
> 
> I created a patch [here](https://www.drupal.org/project/drupal/issues/3229493) for that issue, but I think it's stopped working weary

@ndouglas I think it is that or related to it, but your patch might be working on field validation, but failing since I am using entity validation.  Not sure, but a possibility.

swirtSJW commented:
Thanks for the great review @ndouglas.  I addressed all your concerns in this [last commit](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/14233/commits/f8d76b6f15fc2bacfaa0b4cf08cdf4ddc908075e) .

ndouglas reviewed (approved):
❤️

edmund-dunn reviewed (approved):
Nicely done sir! I didn't thnk of this approach, but it is exactly what is needed! Thanks!
